stm32-modules/thermocycler-refresh/firmware/thermal/thermal_peltier_hardware.c

```diff
@@ -0,0 +1,324 @@
+    // and a pulse of the auto-reload-register + 1 will keep it on 100% of the
+    // time.
+    sConfigOC.OCMode = TIM_OCMODE_PWM1;
+    sConfigOC.Pulse = 0;
+    sConfigOC.OCPolarity = TIM_OCPOLARITY_HIGH;
+    sConfigOC.OCNPolarity = TIM_OCNPOLARITY_HIGH;
+    sConfigOC.OCFastMode = TIM_OCFAST_ENABLE;
+    sConfigOC.OCIdleState = TIM_OCIDLESTATE_RESET;
+    sConfigOC.OCNIdleState = TIM_OCNIDLESTATE_RESET;
+    _initialize_peltier(&_peltiers.peltiers[PELTIER_LEFT], &sConfigOC);
+    _initialize_peltier(&_peltiers.peltiers[PELTIER_RIGHT], &sConfigOC);
+    _initialize_peltier(&_peltiers.peltiers[PELTIER_CENTER], &sConfigOC);
+
+    sBreakDeadTimeConfig.OffStateRunMode = TIM_OSSR_DISABLE;
+    sBreakDeadTimeConfig.OffStateIDLEMode = TIM_OSSI_DISABLE;
+    sBreakDeadTimeConfig.LockLevel = TIM_LOCKLEVEL_OFF;
+    sBreakDeadTimeConfig.DeadTime = 0;
+    sBreakDeadTimeConfig.BreakState = TIM_BREAK_DISABLE;
+    sBreakDeadTimeConfig.BreakPolarity = TIM_BREAKPOLARITY_HIGH;
+    sBreakDeadTimeConfig.BreakFilter = 0;
+    sBreakDeadTimeConfig.BreakAFMode = TIM_BREAK_AFMODE_INPUT;
+    sBreakDeadTimeConfig.Break2State = TIM_BREAK2_DISABLE;
+    sBreakDeadTimeConfig.Break2Polarity = TIM_BREAK2POLARITY_HIGH;
+    sBreakDeadTimeConfig.Break2Filter = 0;
+    sBreakDeadTimeConfig.Break2AFMode = TIM_BREAK_AFMODE_INPUT;
+    sBreakDeadTimeConfig.AutomaticOutput = TIM_AUTOMATICOUTPUT_DISABLE;
+    hal_ret = HAL_TIMEx_ConfigBreakDeadTime(&_peltiers.timer, &sBreakDeadTimeConfig);
+
+    configASSERT(hal_ret == HAL_OK);
+    
+    HAL_TIM_MspPostInit(&_peltiers.timer);
+
+    // Initialize enable pin
+    GPIO_InitTypeDef gpio_init = {
+        .Mode = GPIO_MODE_OUTPUT_PP,
+        .Pull = GPIO_NOPULL,
+        .Speed = GPIO_SPEED_FREQ_HIGH,
+        .Pin = PELTIER_ENABLE_PIN
+    };
+    HAL_GPIO_Init(PELTIER_ENABLE_PORT, &gpio_init);
+
+    _peltiers.initialized = true;
+}
+
+void thermal_peltier_set_enable(const bool enable) {
+    if(!_peltiers.initialized) {
+        return;
+    }
+    _peltiers.enabled = enable;
+    GPIO_PinState enable_val = (enable) ? GPIO_PIN_SET : GPIO_PIN_RESET;
+    HAL_GPIO_WritePin(PELTIER_ENABLE_PORT,
+                      PELTIER_ENABLE_PIN,
+                      enable_val);
+    if(!enable) {
+        __HAL_TIM_DISABLE(&_peltiers.timer);
+        // Set all peltiers to 0 power and update their outputs
+        for(size_t i = 0; i < PELTIER_NUMBER; ++i) {
+            Peltier_t *peltier = &_peltiers.peltiers[i];
+            peltier->power = 0.0f;
+            peltier->direction = PELTIER_HEATING;
+            _update_outputs(peltier);
+        }
+    } else {
+        __HAL_TIM_ENABLE(&_peltiers.timer);
+    }
+}
+
+bool thermal_peltier_get_enable(void) {
+    return _peltiers.enabled;
+}
+
+void thermal_peltier_set_power(const PeltierID id, 
+                               double power,
+                               const PeltierDirection direction) {
+    if(!(id < PELTIER_NUMBER)) {
```

Comment:
should probably return some sort of error here, even if it's just a bool that we can put in to a general "something broke in the peltiers" message

Reply:
Added a bool return here + an error code to send over host comms 👍